contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,401 @@
+        uint256 totalPooledBNB;
+        uint256 totalShares;
+        bytes voteAddress;
+        Description description;
+        Commission commission;
+        uint256[20] slots;
+    }
+
+    struct Description {
+        string moniker;
+        string identity;
+        string website;
+        string details;
+    }
+
+    struct Commission {
+        uint256 rate; // the commission rate charged to delegators
+        uint256 maxRate; // maximum commission rate which validator can ever charge
+        uint256 maxChangeRate; // maximum daily increase of the validator commission
+    }
+    //uint256 updateTime;
+
+    enum SlashType {
+        DoubleSign,
+        DownTime,
+        MaliciousVote
+    }
+
+    struct SlashRecord {
+        uint256 slashAmount;
+        uint256 slashHeight;
+        uint256 jailUntil;
+        SlashType slashType;
+    }
+
+    /*----------------- modifiers -----------------*/
+    modifier onlyInitialized() {
+        require(_initialized != 0, "NOT_INITIALIZED");
+        _;
+    }
+
+    modifier validatorExist(address validator) {
+        require(validators[validator].poolModule != address(0), "VALIDATOR_NOT_EXIST");
+        _;
+    }
+
+    /*----------------- init -----------------*/
+    function initialize() public initializer {
+        minSelfDelegation = INIT_MIN_SELF_DELEGATION;
+        minDelegationChange = INIT_MIN_DELEGATION_CHANGE;
+        maxElectedValidators = INIT_MAX_ELECTED_VALIDATORS;
+        downtimeSlashAmount = INIT_DOWNTIME_SLASH_AMOUNT;
+        doubleSignSlashAmount = INIT_DOUBLE_SIGN_SLASH_AMOUNT;
+        downtimeJailTime = INIT_DOWNTIME_JAIL_TIME;
+        doubleSignJailTime = INIT_DOUBLE_SIGN_JAIL_TIME;
+
+        _initialized = 1;
+    }
+
+    /*----------------- external functions -----------------*/
+    function createValidator(bytes calldata voteAddress, bytes calldata blsProof, Commission calldata commission, Description calldata description) external payable onlyInitialized {
+        uint256 delegation = msg.value;
+        require(delegation >= minSelfDelegation, "INVALID_SELF_DELEGATION");
+        address validator = msg.sender;
+        require(validators[validator].poolModule == address(0), "ALREADY_VALIDATOR");
+        require(voteAddressToValidator[voteAddress] == address(0), "DUPLICATE_VOTE_ADDRESS");
+        require(commission.rate <= commission.maxRate, "INVALID_COMMISSION_RATE");
+        require(commission.maxChangeRate <= commission.maxRate, "INVALID_MAX_CHANGE_RATE");
+
+        // check vote address
+        _checkVoteAddress(voteAddress, blsProof);
+        voteAddressToValidator[voteAddress] = validator;
+
+        // deploy stake pool
+        address poolModule = _deployStakePool(validator);
+        totalPooledBNB += delegation;
+
+        Validator storage valInfo = validators[validator];
+        valInfo.consensusAddress = validator;
+        valInfo.poolModule = poolModule;
+        valInfo.totalPooledBNB = delegation;
```

Comment:
I think we'd better not store `totalPooledBNB` and `totalShares`, but get from the stakePool contract if needed, otherwise we need to deal the gap between the data here and the data in stakePool.

Reply:
Done